test/extended/router/grpc-interop.go

```diff
@@ -3,177 +3,202 @@ package router
-			o.Expect(testCases).ShouldNot(o.BeEmpty())
-			sort.Strings(testCases)
 
-			grpcInteropExecInClusterRouteTests(oc, testCases, gRPCInteropTestCaseIterations)
-			grpcInteropExecInClusterServiceTests(oc, testCases, gRPCInteropTestCaseIterations)
-			grpcInteropExecOutOfClusterRouteTests(oc, testCases, gRPCInteropTestCaseIterations)
+			pemCrt, err := certgen.MarshalCertToPEMString(tlsCrtData)
+			o.Expect(err).NotTo(o.HaveOccurred())
+
+			shardFQDN := oc.Namespace() + "." + defaultDomain
+
+			g.By("Creating routes to test for gRPC interoperability")
+			err = oc.Run("new-app").Args("-f", grpcRoutesConfigPath,
+				"-p", "DOMAIN="+shardFQDN,
+				"-p", "TLS_CRT="+pemCrt,
+				"-p", "TLS_KEY="+derKey,
+				"-p", "TYPE="+oc.Namespace()).Execute()
+			o.Expect(err).NotTo(o.HaveOccurred())
+
+			g.By("Creating a test-specific router shard")
+			shardConfigPath, err = shard.DeployNewRouterShard(oc, 10*time.Minute, shard.Config{
+				FixturePath: grpcRouterShardConfigPath,
+				Name:        oc.Namespace(),
+				Domain:      shardFQDN,
+				Type:        oc.Namespace(),
+			})
+			o.Expect(err).NotTo(o.HaveOccurred(), "new router shard did not rollout")
+
+			// Shard is using a namespace selector so
+			// label the test namespace to match.
+			g.By("By labelling the namespace")
+			err = oc.AsAdmin().Run("label").Args("namespace", oc.Namespace(), "type="+oc.Namespace()).Execute()
+			o.Expect(err).NotTo(o.HaveOccurred())
+
+			g.By("Getting LB service")
+			shardService, err := getRouterService(oc, 10*time.Minute, "router-"+oc.Namespace())
+			o.Expect(err).NotTo(o.HaveOccurred())
+			o.Expect(shardService).NotTo(o.BeNil())
+			o.Expect(shardService.Status.LoadBalancer.Ingress).To(o.Not(o.BeEmpty()))
+
+			testCases := []string{
+				"cancel_after_begin",
+				"cancel_after_first_response",
+				"client_streaming",
+				"custom_metadata",
+				"empty_unary",
+				"large_unary",
+				"ping_pong",
+				"server_streaming",
+				"special_status_message",
+				"status_code_and_message",
+				"timeout_on_sleeping_server",
+				"unimplemented_method",
+				"unimplemented_service",
+			}
+
+			for _, routeType := range []routev1.TLSTerminationType{
+				routev1.TLSTerminationEdge,
+				routev1.TLSTerminationReencrypt,
+				routev1.TLSTerminationPassthrough,
+			} {
+				if routeType == routev1.TLSTerminationEdge {
+					e2e.Logf("skipping %v tests - needs https://github.com/openshift/router/pull/104", routeType)
+					continue
+				}
+
+				g.By("Ensuring route is registered in DNS")
+				if len(shardService.Status.LoadBalancer.Ingress[0].IP) > 0 {
+					host := fmt.Sprintf("grpc-interop-%s.%s", routeType, shardFQDN)
+					addrs, err := resolveHostAsAddress(oc, time.Minute, 15*time.Minute, host, shardService.Status.LoadBalancer.Ingress[0].IP)
+					o.Expect(err).NotTo(o.HaveOccurred())
+					o.Expect(addrs).NotTo(o.BeEmpty())
+				} else {
+					addrs, err := resolveHost(oc, time.Minute, 15*time.Minute, shardService.Status.LoadBalancer.Ingress[0].Hostname)
```

Comment:
We only want to continue hitting the route when the route actually resolved to the new LB's IP address. We loop until that condition is met. We could proceed as the test itself also loops but you get a lot of noise logged because it is hitting the default ingress controller and that doesn't have http/2 enabled. As soon as <any-name>.new-domain resolves we know the route hostname will too. Too subtle?